### PR TITLE
feat(drive): add +pull shortcut for one-way Drive → local mirror

### PR DESCRIPTION
**Pull request by fangshuyu-768**
## Summary

Adds `drive +pull`, a one-way Drive → local mirror command. Recursively lists `--folder-token`, downloads each `type=file` entry under the matching relative path in `--local-dir`, and optionally removes local files absent from the remote (mirror semantics).

Output is split into a `summary` (counts per action) plus `items[]` (per-file detail with `rel_path`, `file_token`, `action`, and `error` on failure).

This is the second of three P1 sync-disk commands (4.1 `+status` is in #692; 4.3 `+push` is the natural follow-up).

## Design notes

- **Pagination + recursion.** Standard 200-per-page list of `--folder-token`; subfolders recurse. Subfolder structure is reproduced under `--local-dir`, with parent directories auto-created by `FileIO.Save`.
- **`type=file` only.** Online docs (`docx`/`sheet`/`bitable`/`mindnote`/`slides`) and shortcuts are skipped — there is no equivalent local binary to write back, so they would otherwise pollute `summary.failed` or download empty placeholder bytes.
- **`--if-exists`** controls per-file collision policy: `overwrite` (default) | `skip`. The framework's enum guard rejects any other value. No `keep-both` — callers who want it can rename and re-pull.
- **`--delete-local` is the only destructive flag and is bound to `--yes`** in `Validate`: without `--yes`, the command refuses to start (no listing, no download, nothing). With `--yes`, downloads run first, then `--local-dir` is walked and any regular file not present in the remote map is `os.Remove`'d. This matches the spec doc's "high-risk-write" intent without making the default pull path require confirmation.
- **`--local-dir` is funneled through `validate.SafeLocalFlagPath`** so errors reference `--local-dir` rather than the framework's default `--file`. `FileIO().Stat` then enforces existence and `IsDir`.
- **Scopes:** `drive:drive.metadata:readonly` + `drive:file:download`. The broader `drive:drive` is disabled by enterprise policy in some tenants; this narrower pair was verified end-to-end.
- **Listing helper is duplicated locally** (`drivePullListRemote`) instead of being reused from `drive_status.go` because that change still sits in open PR #692. Once #692 merges, both copies should be lifted into a shared package-level helper. TODO marker left in `drive_pull.go`.

## Output shape

```json
{
  "summary": {
    "downloaded": 0,
    "skipped": 0,
    "failed": 0,
    "deleted_local": 0
  },
  "items": [
    {"rel_path": "...", "file_token": "...", "action": "downloaded"},
    {"rel_path": "...", "file_token": "...", "action": "skipped"},
    {"rel_path": "...", "file_token": "...", "action": "failed", "error": "..."},
    {"rel_path": "...", "action": "deleted_local"},
    {"rel_path": "...", "action": "delete_failed", "error": "..."}
  ]
}
```

### Static checks
- [x] `go build ./...` clean
- [x] `go vet ./...` clean
- [x] `golangci-lint run --new-from-rev=origin/main` — 0 issues (only the same `nolint:forbidigo` justifications used in `+status` for the walker and a single `os.Remove` in `--delete-local`, with comments)
- [x] Skill format check passes
- [x] `go test $(go list ./... | grep -v cli_e2e) -count=1` — all packages green

### Unit tests (`go test ./shortcuts/drive/... -run TestDrivePull`)
- [x] `TestDrivePullDownloadsAndCreatesParents` — happy path with a nested subfolder; verifies `docx` entries are skipped and parent directories are auto-created
- [x] `TestDrivePullSkipsExistingWhenSkipPolicy` — pre-existing local file is preserved verbatim, `summary.skipped` counts it
- [x] `TestDrivePullDeleteLocalRequiresYes` — `--delete-local` without `--yes` rejected upfront
- [x] `TestDrivePullDeletesLocalOnlyFilesWhenYes` — `--delete-local --yes` removes both root-level orphan and orphan in a subdirectory
- [x] `TestDrivePullRejectsAbsoluteLocalDir` — error message references `--local-dir`
- [x] `TestDrivePullRejectsBadIfExistsEnum` — framework enum guard kicks in
- [x] `TestShortcutsIncludesExpectedCommands` updated to require `+pull`

### E2E dry-run tests (`tests/cli_e2e/drive/drive_pull_dryrun_test.go`)
- [x] `TestDrive_PullDryRun` — request shape: GET /open-apis/drive/v1/files + folder_token + description text
- [x] `TestDrive_PullDryRunRejectsAbsoluteLocalDir` — Validate runs under --dry-run; --local-dir surfaced in error
- [x] `TestDrive_PullDryRunRejectsDeleteLocalWithoutYes` — `--delete-local` guard works under dry-run too
- [x] `TestDrive_PullDryRunRejectsMissingFolderToken` — cobra required-flag enforcement

### End-to-end against a real Drive folder (9 type=file entries across 4 levels)

| Step | Command | Result |
|---|---|---|
| 1 | fresh pull (overwrite default) | `downloaded=9 skipped=0 failed=0 deleted_local=0`; full tree replicated locally including `docs/中文文档.md` and `images/logo.bin` (PNG header bytes intact) |
| 2 | re-pull with `--if-exists=skip` | `downloaded=0 skipped=9 failed=0`; existing local content preserved |
| 3 | `--delete-local` without `--yes` | structured `validation` error: `--delete-local requires --yes (high-risk: deletes local files absent from Drive)` |
| 4 | `--delete-local --yes` after seeding `stale-extra.txt` and `orphan-dir/file.txt` | `downloaded=9 deleted_local=2`; both orphans removed, final tree exactly matches remote |

## Summary by CodeRabbit

* **New Features**
  * Added drive +pull: mirrors a Drive folder into a local directory with dry-run, conditional overwrite/skip (--if-exists), safe-path validation, and optional one-way local cleanup (destructive deletes require --yes). Emits per-item actions and aggregate summary counts.

* **Documentation**
  * New user guide and reference documenting usage, flags, OAuth scopes, and path-safety guidance.

* **Tests**
  * Added unit and end-to-end tests for listing/download flows, validation, dry-run, deletion semantics, symlink safety, and edge cases.

### CODE REVIEW

**Comment by fangshuyu-768** [comment]
Pushed two follow-up commits addressing the CodeRabbit review:

**`45fe4e3` — `fix: --delete-local must not unlink local files shadowed by online docs`**
Critical comment was correct. `drivePullListRemote` now returns two views: `files` (rel_path → file_token, type=file only — drives downloads/skip) and `allPaths` (every entry's rel_path regardless of type — consulted by `--delete-local`). An online doc/shortcut at the same rel_path as a local file no longer makes the local file look orphaned. Same commit also routes the local walk and remove through `vfs.ReadDir` + `vfs.Remove` (replaces `filepath.WalkDir` + `os.Remove`), dropping the two `nolint:forbidigo` justifications. The recursive `vfs.ReadDir` walker preserves "do not follow child symlinks" so the canonical-root escape protections from `145c6a2` remain intact. Added `TestDrivePullDeleteLocalPreservesLocalFileShadowedByOnlineDoc` as the direct regression.

**`8e5adfd` — `fix: count +pull delete failures in summary.failed`**
Also correct. Both `delete_failed` branches now increment `failed` so the summary stops misreporting `failed=0` on partial mirrors. Added `TestDrivePullDeleteLocalCountsFailureInSummary` (forces `vfs.Remove` to fail via `chmod 0o555` on the parent dir).

**`tests/cli_e2e/drive/drive_pull_dryrun_test.go` exit-code contract — keeping as-is.**
The 365-day cap example in `drive_search_dryrun_test.go:258` follows `exit=0 + stdout.error` only because that validation runs inside the `DryRun` callback (it calls `common.NewDryRunAPI().Set("error", ...)`). For hard rejections in the `Validate` callback (cobra enum guards, custom rules), the existing convention in this suite is `exit≠0` — see `TestDriveSearchDryRun_RejectsInvalidSort` (search.go:290) and `TestDriveApplyPermissionDryRun_RejectsFullAccess` (drive_apply_permission_dryrun_test.go:186). `+pull`'s `--local-dir` and `--delete-local` checks live in `Validate`, so the existing tests follow the right contract for that hook.

File: tests/cli_e2e/drive/drive_pull_dryrun_test.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	})
+	require.NoError(t, err)
+	if result.ExitCode == 0 {
+		t.Fatalf("absolute --local-dir must be rejected, got exit=0\nstdout:\n%s", result.Stdout)
+	}
+	combined := result.Stdout + "\n" + result.Stderr
+	if !strings.Contains(combined, "--local-dir") {
+		t.Fatalf("expected --local-dir in error, got:\nstdout:\n%s\nstderr:\n%s", result.Stdout, result.Stderr)
+	}
+}
```
[the remark this answers is not in the document]
Comment (fangshuyu-768) [reply tests/cli_e2e/drive/drive_pull_dryrun_test.go:100]: Keeping exit≠0 here intentionally — `+pull` runs its `--local-dir` and `--delete-local` checks in `Validate`, and the existing convention in this suite for hard `Validate` rejects is non-zero exit (see `TestDriveSearchDryRun_RejectsInvalidSort` and `TestDriveApplyPermissionDryRun_RejectsFullAccess`). `drive +search`'s 365-day case follows `exit=0 + stdout.error` only because that check lives in the `DryRun` callback (`common.NewDryRunAPI().Set("error", ...)`), not in `Validate`. Full rationale: https://github.com/larksuite/cli/pull/696#issuecomment-4335484417

File: shortcuts/drive/drive_pull.go [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+		if deleteLocal {
+			localPaths, err := drivePullWalkLocal(localDir)
+			if err != nil {
+				return err
+			}
+			for _, rel := range localPaths {
+				if _, ok := remoteFiles[rel]; ok {
+					continue
+				}
+				target := filepath.Join(localDir, rel)
```
[the remark this answers is not in the document]
Comment (fangshuyu-768) [reply shortcuts/drive/drive_pull.go]: Done in commit \`45fe4e3\` (and walker swapped back to filepath/os in \`c1b0bed\` to satisfy the \`shortcuts-no-vfs\` depguard rule). \`drivePullListRemote\` now returns two views — \`files\` (type=file, drives downloads) and \`allPaths\` (every Drive entry regardless of type, drives the --delete-local check). New regression \`TestDrivePullDeleteLocalPreservesLocalFileShadowedByOnlineDoc\` stages a Drive containing keep.txt (file) plus notes.docx (docx) with the local side holding both, and asserts --delete-local --yes leaves the locally-shadowed docx untouched.

File: shortcuts/drive/drive_pull.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+				// FileIO has no Remove(); the path is bounded inside cwd by
+				// SafeInputPath in Validate, so a bare os.Remove is acceptable
+				// here. Same justification as the walker below.
+				if err := os.Remove(target); err != nil { //nolint:forbidigo // see comment above
```
[the remark this answers is not in the document]
Comment (fangshuyu-768) [reply shortcuts/drive/drive_pull.go]: Going to skip this one — the repo has a depguard rule \`shortcuts-no-vfs\` in \`.golangci.yml\` that explicitly blocks shortcuts from importing \`internal/vfs\`. The intended substitute is \`runtime.FileIO()\` / \`runtime.ValidatePath()\`, but FileIO has no Remove() and no walker today. drive_status.go uses the same fallback (\`os.Remove\` + \`filepath.WalkDir\` behind \`//nolint:forbidigo\` with a justifying comment); +pull now follows that pattern verbatim in commit \`c1b0bed\`. Lint is clean. Once a runtime-level walker / remove lands, both +status and +pull can swap over.

File: tests/cli_e2e/drive/drive_pull_dryrun_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (fangshuyu-768) [reply tests/cli_e2e/drive/drive_pull_dryrun_test.go:100]: Going to skip this one — the contract you cited (\`drive +search\`) has its validation inside the DryRun callback, so dry-run prints an error envelope at \`exit=0\`. \`+pull\` (and \`+status\` PR #692) put the validation in \`Validate\`, which runs **before** \`handleShortcutDryRun\` (see \`shortcuts/common/runner.go\` ~L769-L770). So an invalid flag exits with code 2 and a structured error envelope, not exit 0. Verified locally with the real binary:\n\n\`\`\`\n$ lark-cli drive +pull --local-dir /etc --folder-token fldcnE2E001 --dry-run\n{\"ok\":false,\"identity\":\"user\",\"error\":{\"type\":\"validation\",\"message\":\"--local-dir: ...\"}}\n(exit=2)\n\`\`\`\n\nThe assertions in this test correctly match that contract.

File: shortcuts/drive/drive_pull.go [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+				if _, ok := remoteFiles[rel]; ok {
+					continue
+				}
+				// FileIO has no Remove(); the absolute path comes from
+				// walking safeRoot, which validate.SafeInputPath has
+				// already bounded inside cwd, so a bare os.Remove is
+				// acceptable here.
+				if err := os.Remove(absPath); err != nil { //nolint:forbidigo // see comment above
+					items = append(items, drivePullItem{RelPath: rel, Action: "delete_failed", Error: err.Error()})
+					continue
```
[the remark this answers is not in the document]
Comment (fangshuyu-768) [reply shortcuts/drive/drive_pull.go:247]: Done in commit \`8e5adfd\`. Both \`delete_failed\` branches (the \`filepath.Rel\` error path and the \`os.Remove\` error path) now \`failed++\` after appending the item, so \`summary.failed\` reflects every entry flagged \`delete_failed\` in \`items[]\`. New test \`TestDrivePullDeleteLocalCountsFailureInSummary\` forces the delete to fail by chmod-ing the local dir 0o555 immediately before invocation (and restoring 0o755 in \`t.Cleanup\` so \`t.TempDir\` teardown still works), and asserts both \`summary.failed >= 1\` and a matching \`delete_failed\` entry in \`items[]\`.

File: shortcuts/drive/drive_pull.go [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+	"path/filepath"
+	"sort"
+	"strings"
+
+	larkcore "github.com/larksuite/oapi-sdk-go/v3/core"
+
+	"github.com/larksuite/cli/extension/fileio"
+	"github.com/larksuite/cli/internal/output"
+	"github.com/larksuite/cli/internal/validate"
+	"github.com/larksuite/cli/internal/vfs"
```
[the remark this answers is not in the document]
Comment (fangshuyu-768) [reply shortcuts/drive/drive_pull.go]: Done in commit \`c1b0bed\`. Reverted the vfs.Remove + vfs.ReadDir-based walker back to os.Remove + filepath.WalkDir, both gated by \`//nolint:forbidigo\` with comments explaining the depguard \`shortcuts-no-vfs\` rule. Lint is now clean.

File: shortcuts/drive/drive_pull.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+// It returns two views:
+//   - files:    rel_path → file_token, for entries with type=file. This is
+//     the "downloadable" subset and drives the download/skip loop.
+//   - allPaths: every entry's rel_path regardless of type (file, online doc,
+//     shortcut, …). --delete-local consults this set so that a local file
+//     sitting at the same rel_path as e.g. an online doc is NOT treated as
+//     orphaned and deleted.
+//
+// Subfolders recurse; online docs and shortcuts are not added to files
+// (no equivalent local binary) but are recorded in allPaths.
```
[the remark this answers is not in the document]
Comment (fangshuyu-768) [reply shortcuts/drive/drive_pull.go]: Done in commit \`de5cbce\`. The folder branch in \`drivePullListRemote\` now records the folder's own rel_path into \`allPaths\` before recursing, so a local regular file shadowed by a remote folder of the same name is no longer treated as orphaned. New regression \`TestDrivePullDeleteLocalPreservesLocalFileShadowedByRemoteFolder\` stages a Drive containing a folder \`shadow\` alongside a downloadable file, with a local regular file at the same \`shadow\` name; --delete-local --yes must leave the shadow file untouched.

**Review by wittam-01 — APPROVED** [review]
